### PR TITLE
[1981] add course create endpoint

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context

Providers need to be able to create new courses.

### Changes proposed in this pull request

- Adds a create method on the course controller. 

### Guidance to review

After speaking to a couple of members of the team validations will be coming in a separate PR. 

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/controllers/api/v2/courses_controller.rb [hunk 1]
```diff
@@ -69,6 +69,21 @@ def destroy
         @course.discard
       end
 
+      def create
+        authorize @provider, :can_create_course?
+        return unless new_course_params.values.any?
+
+        course = Course.new
+        course.assign_attributes(new_course_params)
+        course.update(provider: @provider)
```
Comment (richrace) [inline app/controllers/api/v2/courses_controller.rb]: Is there a reason it's not:

```ruby
course = Course.new(new_course_params.merge(provider: @provider))

if course.save 

else
 
end 
```

Comment (davidgisbey) [reply app/controllers/api/v2/courses_controller.rb]: Just my lack of ruby skills 🤦‍♂

**Comment by davidgisbey** [comment]
[Add course create action.](https://trello.com/c/9bGqnLPl/1981-add-course-create-action)

File: app/controllers/api/v2/courses_controller.rb [hunk 2]
```diff
@@ -115,6 +134,10 @@ def update_site_status_vac_statuses
         end
       end
 
+      def providers_course_codes
+        @provider.courses.map(&:course_code)
+      end
```
Comment (richrace) [inline app/controllers/api/v2/courses_controller.rb]: You don't need this method.

File: spec/services/courses/generate_course_code_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+RSpec.describe Courses::GenerateCourseCodeService do
```
Comment (defong) [inline spec/services/courses/generate_course_code_service_spec.rb]:
```suggestion
describe Courses::GenerateCourseCodeService do
```

File: app/services/courses/generate_course_code_service.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    end
+
+  private
+
+    def valid_letters
+      ('A'..'Z').to_a - %w[O I]
+    end
+
+    def valid_number
+      (0..9).to_a.sample.to_s
```
Comment (defong) [inline app/services/courses/generate_course_code_service.rb]: 0..999 with padding?

File: spec/services/courses/generate_unique_course_code_spec.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+require 'rails_helper'
+
+RSpec.describe Courses::GenerateUniqueCourseCodeService do
```
Comment (defong) [inline spec/services/courses/generate_unique_course_code_spec.rb]:
```suggestion
describe Courses::GenerateUniqueCourseCodeService do
```

File: app/services/courses/generate_course_code_service.rb (hunk 4 above)
Comment (tvararu) [reply app/services/courses/generate_course_code_service.rb]: For those wondering how, I googled for it:

```ruby
> 99.to_s.rjust(3, "0")
=> "099"
> "%03d" % 99
>= "099"
```

Comment (defong) [reply app/services/courses/generate_course_code_service.rb]: you beat me

File: spec/services/courses/generate_course_code_service_spec.rb (hunk 3 above)
Comment (richrace) [reply spec/services/courses/generate_course_code_service_spec.rb]: Any reason why dropping the top level `RSpec` it seems inconsistent between other tests.

File: app/services/courses/generate_course_code_service.rb (hunk 4 above)
Comment (davidgisbey) [reply app/services/courses/generate_course_code_service.rb]: Done. Thanks 👍

File: spec/services/courses/generate_course_code_service_spec.rb (hunk 3 above)
Comment (davidgisbey) [reply spec/services/courses/generate_course_code_service_spec.rb]: Discussed

File: spec/services/courses/generate_unique_course_code_spec.rb (hunk 5 above)
Comment (davidgisbey) [reply spec/services/courses/generate_unique_course_code_spec.rb]: Done

File: app/controllers/api/v2/courses_controller.rb (hunk 2 above)
Comment (davidgisbey) [reply app/controllers/api/v2/courses_controller.rb]: Good spot. Done. Thanks

**Pull request merged**